esp/src/eclwatch/PackageMapPartsWidget.js

```diff
@@ -0,0 +1,356 @@
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+############################################################################## */
+define([
+    "dojo/_base/declare",
+    "dojo/_base/lang",
+    "dojo/i18n",
+    "dojo/i18n!./nls/hpcc",
+    "dojo/_base/array",
+    "dojo/on",
+    "dojo/dom",
+    "dojo/dom-form",
+    "dojo/dom-construct",
+    "dojo/dom-class",
+    "dojo/promise/all",
+
+    "dijit/registry",
+    "dijit/form/ToggleButton",
+    "dijit/ToolbarSeparator",
+    "dijit/form/Button",
+    "dijit/form/ValidationTextBox",
+    "dijit/form/Textarea",
+    "dijit/form/TextBox",
+    "dijit/form/CheckBox",
+    "dijit/Dialog",
+
+    "dgrid/selector",
+
+    "hpcc/GridDetailsWidget",
+    "hpcc/WsDFUXref",
+    "hpcc/DelayLoadWidget",
+    "hpcc/ESPUtil",
+    "hpcc/WsPackageMaps",
+    "hpcc/Utility",
+    "hpcc/FilterDropDownWidget",
+    "hpcc/TargetSelectWidget",
+    "hpcc/ECLSourceWidget"
+], function (declare, lang, i18n, nlsHPCC, arrayUtil, on, dom, domForm, domConstruct, domClass, all,
+                registry, ToggleButton, ToolbarSeparator, Button, ValidationTextBox, Textarea, TextBox, CheckBox, Dialog,
+                selector,
+                GridDetailsWidget, WsDFUXref, DelayLoadWidget, ESPUtil, WsPackageMaps, Utility, FilterDropDownWidget, TargetSelectWidget, ECLSourceWidget) {
+    return declare("PackageMapPartsWidget", [GridDetailsWidget], {
+        i18n: nlsHPCC,
+        gridTitle: nlsHPCC.Parts,
+        idProperty: "Part",
+        addPartsDropDown: null,
+        addPartsDropDownLoaded: null,
+
+        init: function (params) {
+            if (this.inherited(arguments))
+                return;
+            this.packageMap = params.packageMap;
+            this._refreshActionState();
+            this.refreshGrid();
+        },
+
+        postCreate: function (args) {
+            this.inherited(arguments);
+            this.addPartsSelect = registry.byId(this.id + "AddPartsSelect");
+        },
+
+        _onRefresh: function (event) {
+            this.refreshGrid();
+        },
+
+        createGrid: function (domID) {
+            var context = this;
+
+            this.openButton = registry.byId(this.id + "Open");
+            this.addPartsDropDown = new FilterDropDownWidget({
+                 id: this.id + "AddParts",
+                 disabled: false,
+                 label: this.i18n.AddPart
+             }).placeAt(this.openButton.domNode, "after");
+             this.getPartDialog = new Dialog({
+                title: this.i18n.GetPart,
+                style: "width: 600px;"
+            });
+             this.addPartsDropDown.on("apply", function (evt) {
+                if (context.addPartsDropDown.filterForm.validate()){
+                    retVal = context.getFilter();
```

Comment:
Is this retVal actually used to return something?  Maybe it could have a more sensible name?